README
                              gdisofs

Usage:
    
    gdisofs [<options>] <gdi_file> <mountpoint> [<FUSE library options>]
    
mounts image, while fusermount shipped with FUSE library can be used to unmount:

    fusermount -u <mountpoint>

fuseiso options are:

    -p              Maintain mountpoint. Create it if it does not exists on
                    start, delete it on exit.

    -n              Do NOT maintain ~/.mtab.fuseiso . This file have format of
                    /etc/mtab and normally stores information about currently
                    mounted iso images.

    -c iocharset    Specify iocharset to use. Joliet filesystem store filenames
                    in unicode and to show them properly they need to be
                    converted to local charset. Default charset is a current
                    locale charset.

gdisofs supports Sega GD-ROM images consisting of a directory containing data
and audio tracks ripped from a GD-ROM and a .gdi file containing metadata
describing how those tracks are layed out on the physical disc.

Currently only Sega Dreamcast games using the 3-track format are supported.


gdisofs is based on code from fuseisofs and the Linux kernel.
It is licensed under the terms of the GNU GPLv2; see COPYING for details.

NEWS
2017-01-18
        * gdisofs fork adds support for Sega Dreamcast gdi images

2006-09-27
        * Added support to maintain ~/.mtab.fuseiso, maintain mount point directory. Added support for iocharset.

2006-09-26
        * Added support to mounting NRG-image files.

2006-09-25
        * Added support to mounting BIN-image files. One-track only for now.


src/isofs.h
/***************************************************************************
 *   Copyright (C) 2005, 2006 by Dmitry Morozhnikov   *
 *   dmiceman@mail.ru   *
 *                                                                         *
 *   Copyright (C) 2017 snickerbockers                                     *
 *   chimerasaurusrex@gmail.com                                            *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#ifndef _ISOFS_H
#define _ISOFS_H

#include <sys/stat.h>
#include <linux/iso_fs.h>
#include <linux/rock.h>

#include "gdi_file.h"

typedef int (*isofs_dir_fill_t) (void *buf, const char *name,
    const struct stat *stat, off_t off);

typedef struct _isofs_context {
    char const *imagefile;
    gdi_info gdi;

    // 1 FILE* per track, unused tracks will be NULL
    FILE **tracks;

    int pd_have_rr; // 1 if primary descriptor have hierarchy with rrip extension
    struct iso_primary_descriptor pd;
    int supplementary; // 1 if supplementary descriptor found and in effect
    struct iso_supplementary_descriptor sd;
    struct iso_directory_record *root;
    int file_offset; // offset to begin of useful data (for .nrg files)
    int id_offset; // offset to CD001 inside file
    size_t block_size; // raw block size
    size_t block_offset; // offset from block start to data
    size_t data_size; // data size inside block 
    int susp; // parse susp entries
    int susp_skip; // skip bytes from susp SP entry
    int joliet_level; // joliet extension level (1, 2 or 3)
    ino_t last_ino;
} isofs_context;

typedef struct _isofs_inode {
    struct iso_directory_record *record;
    struct stat st;
    ino_t st_ino;
    time_t ctime; // cached value from record->date
    char *sl;
    size_t sl_len;
    char *nm;
    size_t nm_len;
    int cl_block;
    int pl_block;
    size_t zf_size;
    size_t real_size;
    char zf_algorithm[2];
    size_t zf_header_size;
    int zf_block_shift;
    int zf_nblocks;
    int *zf_blockptr;
    int PX; // 1 if PX entry found, st in effect
    int SL; // 1 if SL entry found, sl in effect
    int NM; // 1 if NM entry found, nm in effect
    int CL; // 1 if CL found, cl_block in effect
    int PL; // 1 if PL found, pl_block in effect
    int RE;
    int TF; // 1 if TF entry found, st in effect
    int ZF; // 1 if ZF entry found
} isofs_inode;

// borrowed from zisofs-tools
typedef struct _zf_file_header {
  char magic[8];
  char uncompressed_len[4];
  unsigned char header_size;
  unsigned char block_size;
  char reserved[2];
} zf_file_header;

// macros for iso_directory_record->flags
#define ISO_FLAGS_HIDDEN(x) (*((unsigned char *) x) & 1)
#define ISO_FLAGS_DIR(x) (*((unsigned char *) x) & (1 << 1))

// borrowed from linux kernel rock ridge code

#define SIG(A,B) ((A) | ((B) << 8)) /* isonum_721() */

// borrowed from linux kernel isofs code

/* Number conversion inlines, named after the section in ISO 9660
   they correspond to. */

#include <byteswap.h>

static inline int isonum_711(unsigned char *p)
{
    return *(unsigned char *)p;
}
// static inline int isonum_712(char *p)
// {
//     return *(s8 *)p;
// }
static inline unsigned int isonum_721(char *p)
{
#if defined(WORDS_BIGENDIAN)
    return *(unsigned short *)p;
#else
    return bswap_16(*(unsigned short *)p);
#endif
}
// static inline unsigned int isonum_722(char *p)
// {
//     return be16_to_cpu(get_unaligned((__le16 *)p));
// }
static inline unsigned int isonum_723(char *p)
{
    /* Ignore bigendian datum due to broken mastering programs */
#if defined(WORDS_BIGENDIAN)
    return bswap_16(*(unsigned short *)p);
#else
    return *(unsigned short *)p;
#endif
}
static inline unsigned int isonum_731(char *p)
{
#if defined(WORDS_BIGENDIAN)
    return bswap_32(*(unsigned int *)p);
#else
    return *(unsigned int *)p;
#endif
}
// static inline unsigned int isonum_732(char *p)
// {
//     return be32_to_cpu(get_unaligned((__le32 *)p));
// }
static inline unsigned int isonum_733(char *p)
{
    /* Ignore bigendian datum due to broken mastering programs */
#if defined(WORDS_BIGENDIAN)
    return bswap_32(*(unsigned int *)p);
#else
    return *(unsigned int *)p;
#endif
}

int isofs_real_preinit(char const *gdi_path);
void* isofs_real_init();

int isofs_real_opendir(const char *path);
int isofs_real_readdir(const char *path, void *filler_buf, isofs_dir_fill_t filler);
int isofs_real_getattr(const char *path, struct stat *stbuf);
int isofs_real_readlink(const char *path, char *target, size_t size);
int isofs_real_open(const char *path);
int isofs_real_read(const char *path, char *out_buf, size_t size, off_t offset);
int isofs_real_statfs(struct statfs *stbuf);

#endif // _ISOFS_H


src/gdi_file.c
/***************************************************************************
 *   Copyright (C) 2017 snickerbockers                                     *
 *   chimerasaurusrex@gmail.com                                            *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include <err.h>
#include <stdio.h>
#include <stdlib.h>

#include <glib.h>

#include "gdi_file.h"

/* enforce sane limits */
#define MAX_TRACKS 16
#define MAX_TRACK_FIELDS 16

void parse_gdi(gdi_info *outp, char const *path) {
    long file_sz;

    FILE *stream = fopen(path, "r");

    if (fseek(stream, 0, SEEK_END) != 0)
        err(1, "While loading gdi file \"%s\"", path);

    if ((file_sz = ftell(stream)) < 0)
        err(1, "While loading gdi file \"%s\"", path);

    if (fseek(stream, 0, SEEK_SET) != 0)
        err(1, "While loading gdi file \"%s\"", path);

    /* room to append a terminator at the end just in case */
    long buf_sz = file_sz + 1;
    char *buf = (char*)g_malloc(buf_sz);

    if (fread(buf, sizeof(char), file_sz, stream) != file_sz)
        errx(1, "Failed to read from gdi file \"%s\"", path);
    buf[buf_sz - 1] = '\0';

    char **lines = g_strsplit_set(buf, "\n", MAX_TRACKS);

    if (!lines[0])
        errx(1, "Malformed gdi file \"%s\"\n", path);

    unsigned n_tracks = atoi(lines[0]);
    if (n_tracks > MAX_TRACKS) {
        errx(1, "Malformed gdi file \"%s\": too many tracks (%u)\n",
             path, n_tracks);
    }

    gdi_track *tracks = (gdi_track*)g_malloc(sizeof(gdi_track) * n_tracks);

    char *gdi_dir = g_path_get_dirname(path);

    unsigned track_no;
    for (track_no = 1; track_no <= n_tracks; track_no++) {
        gdi_track *cur_track = tracks + (track_no - 1);
        char **track_info = g_strsplit_set(lines[track_no], " \t",
                                           MAX_TRACK_FIELDS);

        /* maybe it's legal for them to be out of order, IDK... */
        if (!(track_info[0] && (atoi(track_info[0]) == track_no))) {
            errx(1, "Malformed gdi file \"%s\": missing track %d\n",
                 path, track_no);
        }

        if (!track_info[1] || !track_info[2] ||
            !track_info[3] || !track_info[4]) {
            errx(1, "Malformed gdi file \"%s\": track %u corrupted\n",
                 path, track_no);
        }

        cur_track->lba_start = atoi(track_info[1]);
        cur_track->ctrl = atoi(track_info[2]);
        cur_track->sector_size = atoi(track_info[3]);
        cur_track->path = g_build_filename(gdi_dir, track_info[4], NULL);

        g_strfreev(track_info);
    }

    outp->n_tracks = n_tracks;
    outp->tracks = tracks;

    g_free(gdi_dir);
}

void cleanup_gdi(gdi_info *info) {
    unsigned track_no;

    for (track_no = 1; track_no <= info->n_tracks; track_no++) {
        gdi_track *cur_track = info->tracks + (track_no - 1);
        g_free(cur_track->path);
    }

    free(info->tracks);
    info->tracks = NULL;
    info->n_tracks = 0;
}


src/gdi_file.h
/***************************************************************************
 *   Copyright (C) 2017 snickerbockers                                     *
 *   chimerasaurusrex@gmail.com                                            *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#ifndef GDI_FILE_H_
#define GDI_FILE_H_

#include <stdint.h>

#define GDI_DATA_TRACK 3

#define GDI_SECONDARY_DATA_TRACK 5

typedef struct _gdi_track {
    char *path;
    unsigned lba_start;    // block address offset
    unsigned ctrl;         // ???
    unsigned sector_size;  // sector size, typically (but not always) 2352
} gdi_track;

typedef struct _gdi_info {
    unsigned n_tracks;

    gdi_track *tracks;
} gdi_info;

void parse_gdi(gdi_info *outp, char const *path);

#endif


src/fuseiso.c
/***************************************************************************
 *   Copyright (C) 2005, 2006 by Dmitry Morozhnikov   *
 *   dmiceman@mail.ru   *
 *                                                                         *
 *   Copyright (C) 2017 snickerbockers                                     *
 *   chimerasaurusrex@gmail.com                                            *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/


#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <errno.h>
#include <unistd.h>
#include <linux/stat.h>
#include <unistd.h>
#include <getopt.h>
#include <mntent.h>
#include <sys/param.h>
#include <pwd.h>

#include <linux/iso_fs.h>

#define FUSE_USE_VERSION 22
#include <fuse.h>

#include <zlib.h>
#include <locale.h>
#include <langinfo.h>

#include "config.h"
#include "isofs.h"
#include "gdi_file.h"

#ifdef __GNUC__
# define UNUSED(x) x __attribute__((unused))
#else
# define UNUSED(x) x
#endif

static char *imagefile = NULL;
static char *mount_point = NULL;

int maintain_mount_point;
int maintain_mtab;
char* iocharset;

char* normalize_name(const char* fname) {
    char* abs_fname = (char *) malloc(PATH_MAX);
    realpath(fname, abs_fname);
    // ignore errors from realpath()
    return abs_fname;
};

int check_mount_point() {
    struct stat st;
    int rc = lstat(mount_point, &st);
    if(rc == -1 && errno == ENOENT) {
        // directory does not exists, createcontext
        rc = mkdir(mount_point, 0777); // let`s underlying filesystem manage permissions
        if(rc != 0) {
            perror("Can`t create mount point");
            return -EIO;
        };
    } else if(rc == -1) {
        perror("Can`t check mount point");
        return -1;
    };
    return 0;
};

void del_mount_point() {
    int rc = rmdir(mount_point);
    if(rc != 0) {
        perror("Can`t delete mount point");
    };
};

char* get_mtab_path() {
    char* mtab_path = (char*) malloc(PATH_MAX);
    uid_t uid = getuid();
    struct passwd* passwd = getpwuid(uid);
    if(!passwd) {
        fprintf(stderr, "Can`t get home directory for user %d: %s\n", uid, strerror(errno));
        return NULL;
    };
    mtab_path[0] = 0;
    if(passwd->pw_dir) { // may be NULL, who know..
        strncpy(mtab_path, passwd->pw_dir, PATH_MAX - 16);
        mtab_path[PATH_MAX - 1] = 0;
    };
    strcat(mtab_path, "/.mtab.fuseiso");
    return mtab_path;
};

int add_mtab_record() {
    int rc;
    char* mtab_path = get_mtab_path();
    if(!mtab_path) {
        return -EIO;
    };
    int fd = open(mtab_path, O_RDWR | O_CREAT, 0644);
    if(fd < 0) {
        perror("Can`t open mtab");
        return -EIO;
    };
    rc = lockf(fd, F_LOCK, 0);
    if(rc != 0) {
        perror("Can`t lock mtab");
        return -EIO;
    };
    FILE* mtab = setmntent(mtab_path, "a");
    if(!mtab) {
        perror("Can`t open mtab");
        return -EIO;
    };
    struct mntent ent;
    ent.mnt_fsname = imagefile;
    ent.mnt_dir = mount_point;
    ent.mnt_type = "fuseiso";
    ent.mnt_opts = "defaults";
    ent.mnt_freq = 0;
    ent.mnt_passno = 0;    
    rc = addmntent(mtab, &ent);
    if(rc != 0) {
        perror("Can`t add mtab entry");
        return -EIO;
    };
    endmntent(mtab);
    rc = lockf(fd, F_ULOCK, 0);
    if(rc != 0) {
        perror("Can`t unlock mtab");
        return -EIO;
    };
    close(fd);
    free(mtab_path);
    return 0;
};

int del_mtab_record() {
    int rc;
    char* mtab_path = get_mtab_path();
    char new_mtab_path[PATH_MAX];
    if(!mtab_path) {
        return -EIO;
    };
    int fd = open(mtab_path, O_RDWR | O_CREAT, 0644);
    if(fd < 0) {
        perror("Can`t open mtab");
        return -EIO;
    };
    rc = lockf(fd, F_LOCK, 0);
    if(rc != 0) {
        perror("Can`t lock mtab");
        return -EIO;
    };
    strncpy(new_mtab_path, mtab_path, PATH_MAX - 16);
    new_mtab_path[PATH_MAX - 1] = 0;
    strcat(new_mtab_path, ".new");
    FILE* mtab = setmntent(mtab_path, "r");
    if(!mtab) {
        perror("Can`t open mtab");
        return -EIO;
    };
    FILE* new_mtab = setmntent(new_mtab_path, "a+");
    if(!new_mtab) {
        perror("Can`t open new mtab");
        return -EIO;
    };
    struct mntent* ent;
    while((ent = getmntent(mtab)) != NULL) {
        if(strcmp(ent->mnt_fsname, imagefile) == 0 && 
            strcmp(ent->mnt_dir, mount_point) == 0 &&
            strcmp(ent->mnt_type, "fuseiso") == 0) {
            // skip
        } else {
            rc = addmntent(new_mtab, ent);
            if(rc != 0) {
                perror("Can`t add mtab entry");
                return -EIO;
            };
        };
    };
    endmntent(mtab);
    endmntent(new_mtab);
    rc = rename(new_mtab_path, mtab_path);
    if(rc != 0) {
        perror("Can`t rewrite mtab");
        return -EIO;
    };
    rc = lockf(fd, F_ULOCK, 0);
    if(rc != 0) {
        perror("Can`t unlock mtab");
        return -EIO;
    };
    close(fd);
    free(mtab_path);
    return 0;
};

static int isofs_getattr(const char *path, struct stat *stbuf)
{
    return isofs_real_getattr(path, stbuf);
}

static int isofs_readlink(const char *path, char *target, size_t size) {
    return isofs_real_readlink(path, target, size);
};

static int isofs_open(const char *path, struct fuse_file_info *UNUSED(fi))
{
    return isofs_real_open(path);
}

static int isofs_read(const char *path, char *buf, size_t size,
                     off_t offset, struct fuse_file_info *UNUSED(fi))
{
    return isofs_real_read(path, buf, size, offset);
}

static int isofs_flush(const char *UNUSED(path), struct fuse_file_info *UNUSED(fi)) {
    return 0;
};

static void* isofs_init() {
    int rc;
    if(maintain_mtab) {
        rc = add_mtab_record();
        if(rc != 0) {
            exit(EXIT_FAILURE);
        };
    };
    return isofs_real_init();
};

static void isofs_destroy(void* param) {
    if(maintain_mtab) {
        del_mtab_record();
    };
    return;
};

static int isofs_opendir(const char *path, struct fuse_file_info *UNUSED(fi)) {
    return isofs_real_opendir(path);
};

static int isofs_readdir(const char *path, void *buf, fuse_fill_dir_t filler, off_t UNUSED(offset),
    struct fuse_file_info *UNUSED(fi)) {
    return isofs_real_readdir(path, buf, filler);
};

static int isofs_statfs(const char *UNUSED(path), struct statfs *stbuf)
{
    return isofs_real_statfs(stbuf);
}

static struct fuse_operations isofs_oper = {
    .getattr    = isofs_getattr,
    .readlink   = isofs_readlink,
    .open       = isofs_open,
    .read       = isofs_read,
    .flush      = isofs_flush,
    .init       = isofs_init,
    .destroy    = isofs_destroy,
    .opendir    = isofs_opendir,
    .readdir    = isofs_readdir,
    .statfs     = isofs_statfs,
};

void usage(const char* prog) {
    printf("Version: %s\nUsage: %s [-n] [-p] [-c <iocharset>] [-h] <isofs_image_file> <mount_point> [<FUSE library options>]\n"
        "Where options are:\n"
        "    -n                 -- do NOT maintain file $HOME/.mtab.fuseiso\n"
        "    -p                 -- maintain mount point; \n"
        "                          create it if it does not exists and delete it on exit\n"
        "    -c <iocharset>     -- specify iocharset for Joliet filesystem\n"
        "    -h                 -- print this screen\n"
        "\nCommon FUSE library options are:\n"
        "    -f                 -- run in foreground, do not daemonize\n"
        "    -d                 -- run in foreground and print debug information\n"
        "    -s                 -- run single-threaded\n"
        "\nPlease consult with FUSE ducumentation for more information\n",
        VERSION_STR,
        prog);
};

int main(int argc, char *argv[])
{
    setlocale(LC_ALL, ""); // set current locale for proper iocharset
    
    // defaults
    maintain_mount_point = 0;
    maintain_mtab = 1;
    iocharset = NULL;
    
    char c;
    while((c = getopt(argc, argv, "+npc:h")) > 0) {
        switch(c) {
            case 'n':
                maintain_mtab = 0;
                break;
            case 'p':
                maintain_mount_point = 1;
                break;
            case 'c':
                if(optarg) {
                    iocharset = optarg;
                };
                break;
            case 'h':
                usage(argv[0]);
                exit(0);
                break;
            case '?':
            case ':':
                usage(argv[0]);
                exit(EXIT_FAILURE);
                break;
        };
    };
    
    if((argc - optind) < 2) {
        usage(argv[0]);
        exit(EXIT_FAILURE);
    };
    
    imagefile = normalize_name(argv[optind]);

    mount_point = normalize_name(argv[optind + 1]);
    
    // with space for possible -o use_ino arguments
    char **nargv = (char **) malloc((argc + 2) * sizeof(char *));
    int nargc = argc - optind;
    
    nargv[0] = argv[0];
    
    int i;
    int next_opt = 0;
    int use_ino_found = 0;
    for(i = 0; i < nargc - 1; ++i) {
        if(next_opt && !use_ino_found) {
            if(strstr(argv[i + optind + 1], "use_ino")) { // ok, already there
                use_ino_found = 1;
                nargv[i + 1] = argv[i + optind + 1];
            } else { // add it
                char* str = (char*) malloc(strlen(argv[i + optind + 1]) + 10);
                strcpy(str, argv[i + optind + 1]);
                strcat(str, ",use_ino");
                nargv[i + 1] = str;
                use_ino_found = 1;
            };
        } else {
            nargv[i + 1] = argv[i + optind + 1];
        };
        // check if this is -o string mean that next argument should be options string
        if(i > 1 && nargv[i + 1][0] == '-' && nargv[i + 1][1] == 'o') {
            next_opt = 1;
        };
    };
    if(!use_ino_found) {
        nargv[nargc] = "-o";
        nargc++;
        nargv[nargc] = "use_ino";
        nargc++;
    };
    
    if(!iocharset) {
        char *nlcharset = nl_langinfo(CODESET);
        if(nlcharset) {
            iocharset = (char *) malloc(strlen(nlcharset) + 9);
            strcpy(iocharset, nlcharset);
            strcat(iocharset, "//IGNORE");
        };
        if(!iocharset) {
            iocharset = "UTF-8//IGNORE";
        };
    };
    
    int rc;
    if(maintain_mount_point) {
        rc = check_mount_point();
        if(rc != 0) {
            exit(EXIT_FAILURE);
        };
    };
    if(maintain_mount_point) {
        rc = atexit(del_mount_point);
        if(rc != 0) {
            fprintf(stderr, "Can`t set exit function\n");
            exit(EXIT_FAILURE);
        }
    };
    
    // will exit in case of failure
    rc = isofs_real_preinit(imagefile);
    
    return fuse_main(nargc, nargv, &isofs_oper);
};
